FirebaseDatabase/Sources/Utilities/FNextPushId.m

```diff
@@ -93,19 +100,91 @@ + (NSString *)successor:(NSString *_Nonnull)key {
-        stringWithFormat:@"%C", [PUSH_CHARS characterAtIndex:sourceIndex + 1]];
+    unichar character = [next characterAtIndex:i];
+    unichar plusOne = character + 1;
+    BOOL removePreviousCharacter = NO;
+    BOOL replaceWithLowestSurrogatePair = NO;
+    switch (plusOne) {
+    case 0x23: // '#'
+    case 0x24: // '$'
+        plusOne = 0x25;
+        break;
+
+    case 0x2E: // '.'
+    case 0x2F: // '/'
+        plusOne = 0x30;
+        break;
+
+    case 0x5B: // '['
+        plusOne = 0x5C;
+        break;
+
+    case 0x5D: // ']'
+        plusOne = 0x5E;
+        break;
 
-    [next replaceCharactersInRange:NSMakeRange(i, i + 1)
+    case 0x7F: // control character: del
+        plusOne = 0x80;
+        break;
+
+    case HIGH_SURROGATE_PAIR_START: // 0xD800
+        // We added one to 0xD7FF and entered surrogate pair zone
+        // Replace with the lowest surrogate pair here
+        replaceWithLowestSurrogatePair = YES;
+
+    case LOW_SURROGATE_PAIR_END + 1: // 0xE000
+        // If the previous character is the highest surrogate value
+        // then we increment to the value 0xE000 by replacing the surrogate
+        // pair by the single value 0xE000 (the value of plusOne)
+        // Otherwise we increment the high surrogate value and set the low
+        // surrogate value to the lowest.
+        if (i == 0) {
+            // Error, encountered low surrogate without matching high surrogate
+        } else {
+            unichar high = [next characterAtIndex:i - 1];
+            if (high == HIGH_SURROGATE_PAIR_END) { /* highest value for the high
+                                                      part of the pair */
+                // Replace pair with 0xE000 (the value of plusOne)
+                removePreviousCharacter = YES;
+            } else {
+                high += 1;
+                NSString *highStr = [NSString stringWithFormat:@"%C", high];
+
+                [next replaceCharactersInRange:NSMakeRange(i - 1, i)
+                                    withString:highStr];
+                plusOne = LOW_SURROGATE_PAIR_START; /* lowest value for the low
+                                                       part of the pair */
+            }
+        }
+        break;
+    }
+
+    NSString *sourcePlusOne =
+        replaceWithLowestSurrogatePair
+            ? [NSString stringWithFormat:@"%C%C", HIGH_SURROGATE_PAIR_START,
+                                         LOW_SURROGATE_PAIR_START]
+            : [NSString stringWithFormat:@"%C", plusOne];
+
+    NSInteger replaceLocation = i;
+    NSInteger replaceLength = 1;
+    if (removePreviousCharacter) {
+        replaceLocation--;
+        replaceLength++;
+    }
+
+    [next replaceCharactersInRange:NSMakeRange(replaceLocation, replaceLength)
                         withString:sourcePlusOne];
-    return [next substringWithRange:NSMakeRange(0, i + 1)];
+    NSInteger length = i + 1;
+    if (removePreviousCharacter) {
+        length--;
```

Comment:
Google-style usually suggest to use prefix operators where possible: `--length`

Reply:
Will fix.